linalg/tensor.hpp

```diff
@@ -0,0 +1,2055 @@
+          typename = typename std::enable_if<std::is_arithmetic<S>::value ||
+                                             is_dual_number<S>::value>::type>
+MFEM_HOST_DEVICE auto operator*(const tensor<T, n...>& A, S scale) ->
+tensor<decltype(T {} * S{}), n...>
+{
+   tensor<decltype(T{} * S{}), n...> C{};
+   for (int i = 0; i < tensor<T, n...>::first_dim; i++)
+   {
+      C[i] = A[i] * scale;
+   }
+   return C;
+}
+
+/**
+ * @brief divide a scalar by each element in a tensor
+ * @tparam S the scalar value type. Must be arithmetic (e.g. float, double, int) or a dual number
+ * @tparam T the underlying type of the tensor (righthand) argument
+ * @tparam n integers describing the tensor shape
+ * @param[in] scale The numerator
+ * @param[in] A The tensor of denominators
+ */
+template <typename S, typename T, int... n,
+          typename = typename std::enable_if<std::is_arithmetic<S>::value ||
+                                             is_dual_number<S>::value>::type>
+MFEM_HOST_DEVICE auto operator/(S scale, const tensor<T, n...>& A) ->
+tensor<decltype(S {} * T{}), n...>
+{
+   tensor<decltype(S{} * T{}), n...> C{};
+   for (int i = 0; i < tensor<T, n...>::first_dim; i++)
+   {
+      C[i] = scale / A[i];
+   }
+   return C;
+}
+
+/**
+ * @brief divide a tensor by a scalar
+ * @tparam S the scalar value type. Must be arithmetic (e.g. float, double, int) or a dual number
+ * @tparam T the underlying type of the tensor (righthand) argument
+ * @tparam n integers describing the tensor shape
+ * @param[in] A The tensor of numerators
+ * @param[in] scale The denominator
+ */
+template <typename S, typename T, int... n,
+          typename = typename std::enable_if<std::is_arithmetic<S>::value ||
+                                             is_dual_number<S>::value>::type>
+MFEM_HOST_DEVICE auto operator/(const tensor<T, n...>& A, S scale) ->
+tensor<decltype(T {} * S{}), n...>
+{
+   tensor<decltype(T{} * S{}), n...> C{};
+   for (int i = 0; i < tensor<T, n...>::first_dim; i++)
+   {
+      C[i] = A[i] / scale;
+   }
+   return C;
+}
+
+/**
+ * @brief compound assignment (+) on tensors
+ * @tparam S the underlying type of the tensor (lefthand) argument
+ * @tparam T the underlying type of the tensor (righthand) argument
+ * @tparam n integers describing the tensor shape
+ * @param[in] A The lefthand tensor
+ * @param[in] B The righthand tensor
+ */
+template <typename S, typename T, int... n> MFEM_HOST_DEVICE
+tensor<S, n...>& operator+=(tensor<S, n...>& A,
+                            const tensor<T, n...>& B)
+{
+   for (int i = 0; i < tensor<S, n...>::first_dim; i++)
+   {
+      A[i] += B[i];
+   }
+   return A;
+}
+
+/**
+ * @brief compound assignment (+) on tensors
+ * @tparam T the underlying type of the tensor argument
+ * @param[in] A The lefthand tensor
+ * @param[in] B The righthand tensor
+ */
+template <typename T> MFEM_HOST_DEVICE
+tensor<T>& operator+=(tensor<T>& A, const T& B)
+{
+   return A.values += B;
```

Comment:
This does not make sense: `A.values` is an array, you need a loop.

Reply:
Actually, this should be okay. The following functions is where `values` is an array.